cmd/soroban-rpc/lib/preflight/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+extern crate libc;
+extern crate soroban_env_host;
+
+use soroban_env_host::budget::Budget;
+use soroban_env_host::storage::{self, AccessType, SnapshotSource, Storage};
+use soroban_env_host::xdr::{
+    self, AccountId, HostFunction, LedgerEntry, LedgerKey, ReadXdr, ScHostStorageErrorCode,
+    WriteXdr,
+};
+use soroban_env_host::{Host, HostError, LedgerInfo};
+use std::convert::TryInto;
+use std::error;
+use std::ffi::{CStr, CString};
+use std::panic;
+use std::ptr::null_mut;
+use std::rc::Rc;
+use xdr::LedgerFootprint;
+
+// TODO: we may want to pass callbacks instead of using global functions
+extern "C" {
+    // LedgerKey XDR in base64 string to LedgerEntry XDR in base64 string
+    fn SnapshotSourceGet(
+        handle: libc::uintptr_t,
+        ledger_key: *const libc::c_char,
+    ) -> *const libc::c_char;
+    // LedgerKey XDR in base64 string to bool
+    fn SnapshotSourceHas(handle: libc::uintptr_t, ledger_key: *const libc::c_char) -> libc::c_int;
+    // Free Strings returned from Go functions
+    fn FreeGoCString(str: *const libc::c_char);
+}
+
+struct CSnapshotSource {
+    handle: libc::uintptr_t,
+}
+
+impl SnapshotSource for CSnapshotSource {
+    fn get(&self, key: &LedgerKey) -> Result<LedgerEntry, HostError> {
+        let key_xdr = key.to_xdr_base64().unwrap();
```

Comment:
@graydon @jayz22 What's the right way for code like the code here to create a `HostError`? As far as I can tell creating a `HostError` is really difficult unless you have a borrow of a `Host` and you can use it's helper functions to construct the right value. In this code there's no `Host` available for the code to use to construct an error, and so we're ending up with a lot of panics instead.

How do you recommend we map errors in this code?

Reply:
Yeah, but unfortunately `ScStatus` isn't expressive enough.